docs/guides/validation-user-guide.md

```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
+mathematically, there is no hard upper limit which a field value can reach.
+
+Usage of the double dot separator ("`..`") between the bounds is necessary. 
+
+<p class="note">
+In some languages, Protobuf unsigned integers are represented by signed language primitives.
+For example, in Java, a `uint64` is represented with a `long`. If a value of a field in Java will
+overflow into `long` negatives, it will be considered a negative by the validation library. Keep
+that in mind when defining lower bounds.
+</p>
+
+### Regular expressions
+
+For `string` fields, the library provides the `(pattern)` option. Users can define a regular
+expression to match the field values. Also, some common pattern modifiers are available:
+ - `dot_all` (a.k.a. "single line") — enables the dot (`.`) symbol to match all the characters,
+   including line breaks;
+ - `case_insensitive` — allows to ignore the case of the matched symbols;
+ - `multiline` — enables the `^` (caret) and `$` (dollar) signs to match a start and an end of
+   a line instead of a start and an end of the whole expression;
+ - `unicode` — enables matching the whole UTF-8 sequences;
+ - `partial_match` — allows the matched strings to contain a full match to the pattern and some 
+   other characters as well. By default, a string only matches a pattern if it is a full match,
+   i.e. there are no unaccounted for leading and/or trailing characters.
+   
+Example:
+
+```proto
+import "spine/options.proto";
+
+// A link to an HTTP(S) resource.
+message HyperReference {
+    string url = 1 [(pattern) = { 
+            regex: "https?://.+\\..+" 
+            modifier: {
+                case_insensitive: true
+            }
+    }];
+}
+```
+
+It is recommended to use simple patterns due to performance considerations. For example, fully
+fledged URL and email patterns are famously too long to be used in most cases. Treat `(pattern)`
+checks as if they were yet another code with regex matching in it.
+
+### Temporal constraint
+
+Spine provides an option for validating time-bearing types. Those are:
+ - `google.protobuf.Timestamp`;
+ - `spine.time.YearMonth`;
+ - `spine.time.LocalDate`;
+ - `spine.time.LocalDateTime`;
+ - `spine.time.OffsetDateTime`;
+ - `spine.time.ZonedDateTime`;
+ - any user-defined type which implements the Temporal interface (`io.spine.time.Temporal` for
+   Java).
+
+Using the option `(when)`, you may declare that the timestamp should lie in past or in future.
+
+```proto
+import "spine/time_options.proto";
+import "spine/time/time.proto";
+
+// A command to place an order.
+message PlaceOrder {
+    
+    // ...
+
+    spine.time.ZonedDateTime when_placed = 12 [(when).in = PAST];
+    spine.time.ZonedDateTime when_expires = 13 [(when).in = FUTURE];
+}
+```
+
+Note that the value is checked in relation to the current server time. In most cases, this should
+not be an issue. However, be aware that using `FUTURE` in Events and entity states may cause
```

Comment:
This is an absolutely required thing to tell. However, I don't understand why it's bad for immutable Events. And I would tell more about the entity states, as, again, there are some missing pieces for readers between the statement and the consequences.